### PR TITLE
Circle and StraightLine

### PR DESCRIPTION
**Pull request by AQuentin2033**
Starting to address issue #30.  First 2 objects Circle and StraightLine.

I also took the opportunity to clean up some of the missing comments and test errors (duplicated tests ignored by nunit)

As an open question for the future - would there be any objection to adding a reference to Mathnet.Symbolics ?  I would like to add some equation support to the geometric shapes (particularly lines)

### CODE REVIEW

File: src/Spatial/Euclidean/Circle.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+{
+    /// <summary>
+    /// Describes a standard 2 dimentional circle
+    /// </summary>
+    public class Circle : IEquatable<Circle>
+    {
+        /// <summary>
+        /// The center point of the circle
+        /// </summary>
+        public Point2D Center { get; }
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Circle.cs]: Minor style nag: we use style cop conventions meaning ctor before properties. Not very important but makes future diffs cleaner.

File: src/Spatial/Euclidean/Circle.cs [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace MathNet.Spatial.Euclidean
+{
+    /// <summary>
+    /// Describes a standard 2 dimentional circle
+    /// </summary>
+    public class Circle : IEquatable<Circle>
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Circle.cs]: We assume namespances here, more consistent with the rest of the lib would be Circle2D.

File: src/Spatial/Euclidean/ILine.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        /// <param name="y">An optional y coordinate</param>
+        /// <returns>A list of crossing points</returns>
+        IEnumerable<Point2D> YIntercept(double y = 0);
+
+        /// <summary>
+        /// Determines if a given point is on the line
+        /// </summary>
+        /// <param name="point">The point to be tested</param>
+        /// <returns>True if the point is on the line</returns>
+        bool IsOnLine(Point2D point);
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/ILine.cs]: `line.IsOnLine(p)` will read a little weird. Not sure if `IsThrough` is good. An overload with a tolerance might makes sense.

File: src/Spatial/Euclidean/Point2D.cs [hunk 4]
```diff
@@ -54,36 +73,65 @@ public Point2D(double[] data)
...
+        /// <summary>
+        /// Returns the centeroid or center of mass of any set of points
+        /// </summary>
+        /// <param name="points">a list of points</param>
+        /// <returns>the centeroid point</returns>
         public static Point2D Centroid(params Point2D[] points)
         {
             return new Point2D(
                 points.Average(point => point.X),
                 points.Average(point => point.Y));
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Point2D.cs]: I should fix this to not use linq, very inefficient implementation here.

File: src/Spatial/Euclidean/StraightLine.cs [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+        /// <summary>
+        /// Returns the hashocde for a Straightline
+        /// </summary>
+        /// <returns>A hashcode</returns>
+        public override int GetHashCode()
+        {
+            unchecked
+            {
+                return (Gradient.GetHashCode() * 397) ^ yoffset.GetHashCode();
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/StraightLine.cs]: Why is xoffset not used here? Also yoffset is not `readonly`

File: src/Spatial/Euclidean/StraightLine.cs [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+        /// <param name="p2">A point on the line</param>
+        public StraightLine(Point2D p1, Point2D p2)
+        {
+            if (p1 == p2)
+                throw new ArgumentException("A line cannot be formed from two identical points");
+            Gradient = (p2.Y - p1.Y) / (p2.X - p1.X);
+            setOffset(p1.X, p1.Y);
+        }
+
+        private void setOffset(double x, double y)
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/StraightLine.cs]: I prefer `private static void SetOffset(ref double x, ref double y)`. That allows the fields to be `readonly`.
Rename to `SetOffset`, also private methods last according to stylecop.

File: src/Spatial/Euclidean/Circle.cs (hunk 1 above)
Comment (AQuentin2033) [reply src/Spatial/Euclidean/Circle.cs]: ok, will update styles.  I didn't use readonly public fields deliberately in order to be able to later add support for lazy calculation where multiple construction methods exist

File: src/Spatial/Euclidean/Circle2D.cs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace MathNet.Spatial.Euclidean
+{
+    /// <summary>
+    /// Describes a standard 2 dimentional circle
+    /// </summary>
+    public class Circle2D : IEquatable<Circle2D>
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Circle2D.cs]: Make this a struct.

File: src/Spatial/Euclidean/Circle2D.cs [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+            Radius = radius;
+        }
+
+        /// <summary>
+        /// Creates a Circle from three points which are not colinnear
+        /// </summary>
+        /// <param name="pointA">A point on the circle</param>
+        /// <param name="pointB">A point on the circle</param>
+        /// /// <param name="pointC">A point on the circle</param>
+        public Circle2D(Point2D pointA, Point2D pointB, Point2D pointC)
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Circle2D.cs]: This is perhaps nicer as a factory method: `var circle = Cricle2D.FromThreePoints(p1, p2, p3)` Not sure if `FromThreePoints` is a good name, a bit redundant with three give three args.

File: src/Spatial/Euclidean/Line2D.cs [hunk 9]
```diff
@@ -252,4 +252,4 @@ public override int GetHashCode()
         }
         # endregion
     }
-}
\ No newline at end of file
+}
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/Line2D.cs]: Remove this from the diff.

File: src/Spatial/Euclidean/PlaneGeometry/LineSegment.cs [hunk 10]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+
+namespace MathNet.Spatial.Euclidean.PlaneGeometry
+{
+    public class LineSegment : LineSegment<StraightLine>
```
Comment (JohanLarsson) [inline src/Spatial/Euclidean/PlaneGeometry/LineSegment.cs]: class and should probably be named LineSegment2D

**Comment by AQuentin2033** [comment]
split as requested

**Pull request closed without merging**